### PR TITLE
Add support for ROCK 4 (aka Rockpi4)

### PR DESCRIPTION
**Pull request by jforissier**
Add rockpi4.mk and as well as U-Boot and Linux kernel config fragments plus one TF-A patch (fix console baudrate) and one U-Boot patch (enable OP-TEE in the DT).

Similar to other platforms such as HiKey, typical build commands are:

  $ make toolchains
  $ make -j$(nproc)
  $ make flash

The bootable image is out/rockpi4.img (128 MB). "make flash" transfers it to the onboard eMMC, but it may also be written to a micro SD card. Note however that the board tries to boot from eMMC first, so once "make flash" has been used, the eMMC has to be cleared to be able to boot from micro SD again. This can be done with "make nuke-emmc".

Tested on Rockpi4B.

Link: https://wiki.radxa.com/Rock4
Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

File: rockpi4.mk [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+else
+TF_A_LOGLVL ?= 40
+TF_A_OUT = $(TF_A_PATH)/build/rk3399/debug
+endif
+
+TF_A_FLAGS ?= ARCH=aarch64 PLAT=rk3399 SPD=opteed DEBUG=$(TF_A_DEBUG) \
+	      LOG_LEVEL=$(TF_A_LOGLVL)
+
+$(TF_A_PATH)/.stamp_patched:
+	patch -d $(TF_A_PATH) -p1 <$(ROOT)/build/rockpi4/tfa/0001-rk3399-set-console-baudrate-to-1500000.patch
```
Comment (jbech-linaro) [inline rockpi4.mk:60]: It this something in the works? I.e., looks like this is something that should be sent upstream, or?

File: rockpi4.mk [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+UBOOT_FLAGS ?= CROSS_COMPILE=$(CROSS_COMPILE_NS_KERNEL) \
+	       CC=$(CROSS_COMPILE_NS_KERNEL)gcc \
+	       HOSTCC="$(CCACHE) gcc"
+
+UBOOT_EXPORTS ?= BL31=$(TF_A_OUT)/bl31/bl31.elf TEE=$(OPTEE_OS_BIN)
+
+u-boot-defconfig: $(UBOOT_PATH)/.config
+
+$(UBOOT_PATH)/.stamp_patched:
+	patch -d $(UBOOT_PATH) -p1 <$(ROOT)/build/rockpi4/u-boot/0001-rockpi4-add-OP-TEE-nodes.patch
```
Comment (jbech-linaro) [inline rockpi4.mk:88]: Ditto, upstream the patch at some point?

**Comment by jbech-linaro** [comment]
With or without comments addressed:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: rockpi4.mk (hunk 1 above)
Comment (jenswi-linaro) [reply rockpi4.mk:60]: Hmm, someone has been "tainted" by yocto? ;-)

This must be quite annoying when working on TF-A.

Comment (jforissier) [reply rockpi4.mk:60]: > Hmm, someone has been "tainted" by yocto? ;-)

Nooooo! I swear! :wink: 

> It this something in the works? I.e., looks like this is something that should be sent upstream, or?

Absolutely, I plan on sending a patch, but it may take some time until this gets into a release. Especially since I am not familiar with TF-A upstreaming at all (Gerrit etc.). In the meantime I thought this was an OK-ish solution -- but note that even without this patch things should work, just without the initial TF-A and OP-TEE logs.

File: rockpi4.mk (hunk 2 above)
Comment (jforissier) [reply rockpi4.mk:88]: That one is much more problematic, because I think the node must be present only when OP-TEE is present (same problem as for the kernel, you don't want U-Boot to invoke an SMC if there's no EL3/SEL1).

**Comment by jenswi-linaro** [comment]
I'm not overjoyed with these patches applied on top of gits. I would prefer using forks just as we do with the Linux kernel.

**Comment by jforissier** [comment]
> I'm not overjoyed with these patches applied on top of gits. I would prefer using forks just as we do with the Linux kernel.

You know what, let's drop both patches. The TF-A console patch doesn't bring much, we can leave without the BL* and OP-TEE banners until it's fixed upstream. As for the OP-TEE DT nodes, they don't have to be in U-Boot, because U-Boot doesn't invoke OP-TEE in our dev config; it's good enough to patch the Linux DT which can be done with a proper commit in our fork. I will add a commit to https://github.com/linaro-swg/linux/pull/106.

File: rockpi4.mk [hunk 3]
```diff
@@ -0,0 +1,231 @@
+COMPILE_NS_USER ?= 64
+override COMPILE_NS_KERNEL := 64
+COMPILE_S_USER ?= 64
+COMPILE_S_KERNEL ?= 64
+
+CFG_TEE_CORE_LOG_LEVEL ?= 2
```
Comment (jenswi-linaro) [inline rockpi4.mk:6]: Why is this needed?

File: rockpi4.mk [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+rkdeveloptool-distclean:
+	$(MAKE) -C $(RKDEVELOPTOOL_PATH) clean
+
+clean: rkdeveloptool-clean
+
+$(LOADER_BIN):
+	cd $(BINARIES_PATH) && \
+		wget https://dl.radxa.com/rockpi/images/loader/$(notdir $(LOADER_BIN))
```
Comment (jenswi-linaro) [inline rockpi4.mk:200]: We should probably remove the downloaded file if this fails for some reason, network timeout in the middle for instance.

File: rockpi4.mk [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+nuke-emmc: $(LOADER_BIN) $(RKDEVELOPTOOL_BIN)
+	@echo
+	@echo "** WARNING: this command will make the onboard eMMC unbootable!"
+	@echo "It can be used to boot from the SD card again."
+	$(call flash-help)
+	dd if=/dev/zero of=$(BINARIES_PATH)/zero.img bs=1M count=64
+	$(RKDEVELOPTOOL_BIN) db $(LOADER_BIN)
+	sleep 1
+	$(RKDEVELOPTOOL_BIN) wl 0 $(BINARIES_PATH)/zero.img
```
Comment (jenswi-linaro) [inline rockpi4.mk:231]: Is it possible to persuade `wl` to read from `stdin` instead? That way we wouldn't need a temporary file.

File: rockpi4.mk (hunk 3 above)
Comment (jforissier) [reply rockpi4.mk:6]: It is not. Will remove.

File: rockpi4.mk (hunk 4 above)
Comment (jforissier) [reply rockpi4.mk:200]: I agree. I will add `.DELETE_ON_ERROR` to take care of that.

File: rockpi4.mk (hunk 5 above)
Comment (jforissier) [reply rockpi4.mk:231]: Apparently not. The `rkdeveloptool` command needs a file name, it won't read from stdin. Worse: it needs a seekable file -- it does `lseek(... SEEK_END)` -- so bash process substitution `<()` won't work :-/

Comment (jenswi-linaro) [reply rockpi4.mk:232]: OK, thanks for looking into it.

File: rockpi4.mk [hunk 6]
```diff
@@ -195,6 +193,9 @@ rkdeveloptool-distclean:
 
 clean: rkdeveloptool-clean
 
+# In case wget fails mid-way
+.DELETE_ON_ERROR:
```
Comment (jenswi-linaro) [inline rockpi4.mk:197]: This is a global setting, affecting all targets that can fail. I'm not sure where it's best to put it, but this close to the end of the file should make it hard to notice.

Comment (jforissier) [reply rockpi4.mk:197]: > This is a global setting, affecting all targets that can fail.

Correct.

> I'm not sure where it's best to put it, but this close to the end of the file should make it hard to notice.

I can move it up

**Comment by jforissier** [comment]
Updated

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by TaihuLight** [comment]
@jforissier @jbech-linaro @jenswi-linaro 
 
The newest plat-rockchip: [RK3588](https://www.rutronik.com/fileadmin/Rutronik/Downloads/printmedia/products/01_semiconductors/Rockchip_Flyer_Microprocessor_Highlights_Online.pdf) also support ARM Trustzone. It is interesting and necessary to build and run OP-TEE on the new board [ROCK5](https://wiki.radxa.com/Rock5) with RK3588 for several developers and users of OP-TEE. However, does the current version of OP-TEE support RK3588? How to build it on Rock 5 running Ubuntu 20.04?

The following issues may be helpful?
#https://github.com/OP-TEE/optee_os/issues/4197

[Rockchip_RK3588_Highlights.pdf](https://github.com/OP-TEE/optee_os/files/10128303/Rockchip_RK3588_Highlights.pdf)
